### ISSUE TITLE
Very little match between Seurat HTODemux and DemuxEM results

### ISSUE THREAD

**Issue by heathergeiger** [klarman-cell-observatory/demuxEM#9]
We are currently testing a nuclear hashing protocol on mouse nuclei here at the New York Genome Center. We did an experiment hashing 8 different mouse samples (7 sample different mouse regions, 1 sample PBMCs). Anyway, we were getting a somewhat high number of HTO-negative cells (that could not be tagged for any sample). We thought perhaps a different method could help, and decided to try DemuxEM.

Attached please find the input CITE-Seq-Count matrix (tag x cell) in CSV format. This was obtained by running CITE-Seq-Count on a whitelist of a bit less than 100,000 barcodes. This whitelist was obtained by taking all first 16bp of R1 sequences from the HTO FASTQ to get a list of cell barcodes. We then required that these cell barcodes occur at least 10x (so at least 10 reads) in the HTO library, and that they be included in the 10X whitelist. This was a way we thought of to still include empty droplets in the CITE-Seq-Count output, while still minimizing CITE-Seq-Count runtime and memory usage versus a longer whitelist.

Also attached please find the list of cell barcodes from the "filtered_feature_bc_matrix" from CellRanger. Finally, the DemuxEM results in CSV format, obtained by running the following code.

```
results = pegasusio.read_input("nuclear_hashing_test.out.demuxEM.zarr.zip")
results = results.obs

results.to_csv(path_or_buf="nuclear_hashing_test_default_min_nUMI_results_from_pegasusio.csv")
```

I have not attached the raw h5 file for the RNA as it is too large. But let me know if you have a way for me to provide that and I can if you need it.

This was the code I used to run DemuxEM. Raw H5 file copied from outs/raw_feature_bc_matrix.h5 under CellRanger directory.

```
DemuxEM raw_feature_bc_matrix.h5 nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads.csv nuclear_hashing_test
```

I also tried running Seurat HTODemux on all cells in the filtered RNA output, using the following code:

```
#Transpose file for faster read into R.

tr ',' '\t' < nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads.csv | awk '
{ 
    for (i=1; i<=NF; i++)  {
        a[NR,i] = $i
    }
}
NF>p { p = NF }
END {    
    for(j=1; j<=p; j++) {
        str=a[1,j]
        for(i=2; i<=NR; i++){
            str=str" "a[i,j];
        }
        print str
    }
}' | tr ' ' '\t' > nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads_transposed.txt

echo Barcode > temp1
cut -f 1 nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads_transposed.txt | head -n 96727 >> temp1
cut -f 2- nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads_transposed.txt > temp2
paste temp1 temp2 > nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads_transposed.txt_new
mv nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads_transposed.txt_new nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads_transposed.txt
```

```
library(Seurat)
HTO_matrix <- read.table("nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads_transposed.txt",header=TRUE,row.names=1)
barcodes <- readLines("nuclear_hashing_test_barcodes_from_10X_filtered_output.txt")
barcodes <- intersect(barcodes,rownames(HTO_matrix))
HTO_matrix <- HTO_matrix[barcodes,]
seurat.obj <- CreateSeuratObject(t(as.matrix(HTO_matrix)))
seurat.obj[["HTO"]] <- CreateAssayObject(counts = t(as.matrix(HTO_matrix)))
#Apply a minimal filter of requiring at least 100 total nUMI across all HTOs. Should remove very few cells.
seurat.obj <- subset(seurat.obj,nCount_HTO >= 100)
#Normalize HTO assay, then run HTODemux.
seurat.obj <- NormalizeData(seurat.obj,assay = "HTO",normalization.method = "CLR",margin=1)
seurat.obj <- HTODemux(seurat.obj,assay = "HTO",seed=1392)
#Get count of assignments.
table(seurat.obj$HTO_classification.global)
table(seurat.obj$hash.ID)
#Print to CSV.
to_print <- data.frame(Barcode = rownames(seurat.obj@meta.data),Assignment = seurat.obj$hash.ID)
write.csv(to_print,file="nuclear_hashing_test_seurat_results_vs_cellranger_filtered_barcodes.csv",row.names=FALSE,quote=FALSE)
```

I have included these results in CSV format as well.

I am getting 11,525 singlets according to Seurat for the cells in the filtered RNA output from CellRanger, compared to only 4,679 singlets for these same cells according to DemuxEM. DemuxEM reports a lot of additional singlets among the cells that were excluded from the CellRanger filtered output, which could be interesting. But it is still strange that we get so many fewer singlets among the same set of cells using DemuxEM.

Any help debugging would be much appreciated. Perhaps I just need to play around with some settings of DemuxEM.

I should also note that when we made a heatmap of the Seurat HTODemux results vs. the HTO matrix, the results generally seemed sensible. So I don't think the results should be so drastically different here, considering the Seurat results were generally sensible.
[nuclear_hashing_test_seurat_results_vs_cellranger_filtered_barcodes.txt](https://github.com/klarman-cell-observatory/demuxEM/files/5105693/nuclear_hashing_test_seurat_results_vs_cellranger_filtered_barcodes.txt)
[nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads_transposed.txt](https://github.com/klarman-cell-observatory/demuxEM/files/5105694/nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads_transposed.txt)
[nuclear_hashing_test_barcodes_from_10X_filtered_output.txt](https://github.com/klarman-cell-observatory/demuxEM/files/5105695/nuclear_hashing_test_barcodes_from_10X_filtered_output.txt)
[nuclear_hashing_test_default_min_nUMI_results_from_pegasusio.txt](https://github.com/klarman-cell-observatory/demuxEM/files/5105696/nuclear_hashing_test_default_min_nUMI_results_from_pegasusio.txt)
[nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads.txt](https://github.com/klarman-cell-observatory/demuxEM/files/5105697/nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads.txt)

**Comment by heathergeiger** [comment]
Some files attached with ".txt" suffix otherwise Github would not let me upload.

Also attaching some of the diagnostic plots, as I forgot to mention I ran with --generate-diagnostic-plots option.

[nuclear_hashing_test.rna_demux.hist.pdf](https://github.com/klarman-cell-observatory/demuxEM/files/5105715/nuclear_hashing_test.rna_demux.hist.pdf)
[nuclear_hashing_test.real_content.hist.pdf](https://github.com/klarman-cell-observatory/demuxEM/files/5105716/nuclear_hashing_test.real_content.hist.pdf)
[nuclear_hashing_test.ambient_hashtag.hist.pdf](https://github.com/klarman-cell-observatory/demuxEM/files/5105717/nuclear_hashing_test.ambient_hashtag.hist.pdf)
[nuclear_hashing_test.background_probabilities.bar.pdf](https://github.com/klarman-cell-observatory/demuxEM/files/5105718/nuclear_hashing_test.background_probabilities.bar.pdf)

**Comment by heathergeiger** [comment]
File is definitely well under 1TB so I can upload to DropBox. My email is hmgeiger@nygenome.org.

**Comment by yihming** [comment]
I've shared my Dropbox space with you. You can simply upload your h5 file to the shared folder. Thanks!

**Comment by heathergeiger** [comment]
Done. Let me know if you got it.

**Comment by yihming** [comment]
I've got it. Thanks!

**Comment by yihming** [comment]
When running demuxEM on my machine, it gives 18,559 singlets, 6934 doublets, and 55,465 unknowns.

When running Seurat with your script, it gives 11,525 singlets, 2710 doublets, and 6423 Negatives, which is consistent with your result.

I used demuxEM 0.1.4, with the following command:
```
demuxEM -p 8 --generate-diagnostic-plots nuclear_hashing_raw_feature_bc_matrix.h5 nuclear_hashing_test_CITE-Seq-Count_results_with_whitelist_from_10plus_reads.csv result
```
I also ran the command without `--generate-diagnostic-plots` option, and the result remained the same. The on-screen output is "result.log" file attached, and the cell barcode info is "nuclear_hashing_test_default_min_nUMI_results_from_pegasusio.txt" attached below.
[result.log](https://github.com/klarman-cell-observatory/demuxEM/files/5120670/result.log)
[nuclear_hashing_test_default_min_nUMI_results_from_pegasusio.txt](https://github.com/klarman-cell-observatory/demuxEM/files/5120672/nuclear_hashing_test_default_min_nUMI_results_from_pegasusio.txt)

What I would suggest is: (1) Update demuxEM to version `0.1.4` if you are using an older version; (2) update PegasusIO to version `0.2.5` if you are using an older version. Then try the command again.

Let me know if the issue still exists.

**Comment by yihming** [comment]
One more comment for your information: when comparing the predicted singlets in Seurat and DemuxEM, there are 2558 in common.

**Comment by yihming** [comment]
I just found that "nuclear_hashing_test_default_min_nUMI_results_from_pegasusio.txt" file you attached has 18,559 predicted singlets. So do you mean that only 4679 singlets were predicted from the 10X filtered gene-count matrix, not the raw count?

**Comment by yihming** [comment]
I was able to reproduce your situation on my machine.

In general, it is important to use the **raw** count matrix, not filtered one, in demuxEM, as it uses raw counts to estimate the background. This may explain why you get a lot fewer predicted singlets from 10X filtered gene-count matrix comparing with the raw count.

**Comment by heathergeiger** [comment]
"So do you mean that only 4679 singlets were predicted from the 10X filtered gene-count matrix, not the raw count?"

Sort of? I used the raw count matrix in DemuxEM. This produced a set of assignments for a set of barcodes that included both the filtered barcodes as well as additional barcodes. I then compared the assignments *only for the filtered barcodes*, but from this output based on the raw data, to Seurat run *only for the filtered barcodes*.

I understand that DemuxEM may find additional singlets that would not be found by using only the filtered barcode set. However what I don't understand is why DemuxEM is reporting such a higher proportion of HTO-negative (unknown) assignments within the same set of barcodes compared to Seurat.

Anyway, given this discrepancy, would you recommend just ignoring whether or not a given barcode is in the filtered list from CellRanger and just proceeding on downstream analysis of all 18,559 singlets as classified by DemuxEM?

From a practical point of view I will admit this is not necessarily the most desirable compared to Seurat. The number of genes per cell for cells classified as singlets by Seurat (using the filtered barcode set) is much higher than by DemuxEM (using the raw barcode set). 

![nuclear_hashing_test_nGene_using_seurat_vs_demuxem_singlets](https://user-images.githubusercontent.com/36860218/91110784-d85b8180-e64c-11ea-8c33-b59e2a478e54.png)

But if ambient RNA is an issue in this dataset perhaps this is a misleading comparison? As is though, it seems the fact that so many of the DemuxEM singlets have worse quality RNA than the Seurat singlets seems like a not so great tradeoff, even if more singlets overall are found by DemuxEM over Seurat.

**Comment by heathergeiger** [comment]
Update: Looks like when I input the HTO counts, I did not include a column name in the header for the first column (with the tag names). If you add "Antibody" in the header for this column instead of a blank space, you get much more reasonable results, in line with what we got for Seurat.

Might be good to include a bit more explanation of the format of input_hto_csv_file in the documentation. But other than that, was a pretty simple fix. Thanks!
